Patches/Specular.cpp

```diff
@@ -0,0 +1,396 @@
+	case ModelID::chr_jms_rhhl_jms:
+	case ModelID::chr_jms_rhhh_jms:
+		return true;
+	}
+
+	return false;
+}
+
+static bool IsMariaExcludingEyes(ModelID id)
+{
+	switch (id)
+	{
+	case ModelID::chr_mar_lll_mar:
+	case ModelID::chr_mar_hhh_mar:
+	case ModelID::chr2_mar_lxx_mar:
+	case ModelID::chr_mar_rlll_mar:
+	case ModelID::chr_mar_rhhh_mar:
+	case ModelID::chr2_mar_rlxx_mar:
+		if (GetCurrentMaterialIndex() != 3) return true;
+		break;
+	case ModelID::chr_item_dmr:
+		if (GetCurrentMaterialIndex() != 1) return true;
+		break;
+	}
+
+	return false;
+}
+
+static bool IsMariaEyes(ModelID id)
+{
+	switch (id)
+	{
+	case ModelID::chr_mar_lll_mar:
+	case ModelID::chr_mar_hhh_mar:
+	case ModelID::chr2_mar_lxx_mar:
+	case ModelID::chr_mar_rlll_mar:
+	case ModelID::chr_mar_rhhh_mar:
+	case ModelID::chr2_mar_rlxx_mar:
+		if (GetCurrentMaterialIndex() == 3) return true;
+		break;
+	case ModelID::chr_item_dmr:
+		if (GetCurrentMaterialIndex() == 1) return true;
+		break;
+	}
+
+	return false;
+}
+
+static void __cdecl Part0(ModelOffsetTable* pOffsetTable, void* arg2)
+{
+	// This function replaces a call to `void DoActorOpaqueStuff(ModelOffsetTable* pOffsetTable, void* arg2)`
+	// Backup the materialCount and pointer to first material for use in later Parts
+
+	materialCount = pOffsetTable->materialCount;
+	pFirstMaterial = (ModelMaterial*)((char*)pOffsetTable + pOffsetTable->materialsOffset);
+
+	DoActorOpaqueStuff(pOffsetTable, arg2);
+}
+
+static int __cdecl Part1()
+{
+	// This function replaces a call to `int GetLightSourceCount()`
+	// When no D3D_DIRECTIONAL light sources exist, set our booleans and return 1 greater than reality
+
+	inSpecialLightZone = false;
+	useFakeLight = true;
+	fakeLightIndex = -1;
+	pCurrentMaterial = nullptr;
+
+	int lightSourceCount = GetLightSourceCount();
+
+	for (int i = 0; i < lightSourceCount; i++)
+	{
+		auto pLight = GetLightSourceStruct(i);
+		if (pLight->light.Type == D3DLIGHT_DIRECTIONAL)
+			useFakeLight = false;
+
+		if (pLight->light.Type == D3DLIGHT_SPOT)
+			inSpecialLightZone = true;
+	}
+
+	if (useFakeLight)
+	{
+		fakeLightIndex = lightSourceCount;
+		return lightSourceCount + 1;
+	}
+
+	return lightSourceCount;
+}
+
+static LightSource* __cdecl Part2(int index)
+{
+	// This function replaces a call to `LightSourceStruct* GetLightSourceStruct(int index)`
+	// When we hit the index 1 greater than the real light source count, return our fake
+
+	if (useFakeLight && index == fakeLightIndex)
+		return &fakeLight;
+	else
+		return GetLightSourceStruct(index);
+}
+
+static void Part3(ModelMaterial* pModelMaterial)
+{
+	// This function replaces a call to `void ActorDrawOpaque(ModelMaterial* pModelMaterial)`
+	// We copy off the pointer to a static variable, so we can reference it in Part 4
+
+	pCurrentMaterial = pModelMaterial;
+	ActorDrawOpaque(pModelMaterial);
+}
+
+static HRESULT __stdcall Part4(IDirect3DDevice8* /*This*/, DWORD Register, void* pConstantData, DWORD ConstantCount)
```

Comment:
Is there a reason this function is not in the "IDirect3DDevice8.cpp" file?  It should be part of the `SetPixelShaderConstant()` function in that file.

Reply:
Also can we get a better name for this?

Reply:
Good catch! I've moved this into `IDirect3DDevice8.cpp`. I now set `specularFlag` to 2 in `HookActorOpaqueDraw` so it doesn't trigger erroneously. This does add a tiny bit of complexity as there is an earlier call to `SetPixelShaderConstant()` that we don't want to touch, so we have to 'tick' `specularFlag` down and only run the code when it's equal to 1.

Reply:
Thanks.  It looks like it will always be the second call after `HookActorOpaqueDraw()`.